### PR TITLE
Adds resource enrichment

### PR DESCRIPTION
**Pull request by Harnoor-se7en**
Adds ResourceAttributeEnricher to attach infra metadata to span tags.

### CODE REVIEW

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  public void init(Config enricherConfig, ClientRegistry clientRegistry) {
+    if (enricherConfig.hasPath(RESOURCE_ATTRIBUTES_CONFIG_KEY)) {
+      resourceAttributesToAdd = enricherConfig.getStringList(RESOURCE_ATTRIBUTES_CONFIG_KEY);
+    }
+  }
+
+  @Override
+  public void enrichEvent(StructuredTrace trace, Event event) {
+    try {
+      if (resourceAttributesToAdd == null || resourceAttributesToAdd.isEmpty()) {
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Why will this ever be null?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        return;
+      }
+      if (event.getResourceIndex() < 0) {
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Add all of these checks to a method `validatePreconditions`?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      }
+      if (event.getAttributes() == null) {
+        return;
+      }
+      Map<String, AttributeValue> attributeMap = event.getAttributes().getAttributeMap();
+      if (attributeMap == null) {
+        return;
+      }
+      for (String resourceAttributeKey : resourceAttributesToAdd) {
+        Optional<AttributeValue> resourceAttribute =
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Rename to `resourceAttributeMaybe`

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            getResourceAttribute(trace, event, resourceAttributeKey);
+        resourceAttribute.ifPresent(
+            attributeValue -> attributeMap.putIfAbsent(resourceAttributeKey, attributeValue));
+      }
+    } catch (Exception e) {
+      LOGGER.error("Exception while enriching event with resource attributes.", e);
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Log some span metadata?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java [hunk 5]
```diff
@@ -39,6 +39,24 @@ public static Map<String, AttributeValue> getAttributesForFirstExistingKey(
     return Collections.unmodifiableMap(attributes);
   }
 
+  public static Optional<org.hypertrace.core.datamodel.AttributeValue> getAttribute(
+      Attributes attributes, String key) {
+    return Optional.ofNullable(attributes)
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java:44]: Use filter with findFirst

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+  @Test
+  public void noAttributes() {
+    Event event = mock(Event.class);
+    when(event.getAttributes()).thenReturn(null);
+    resourceAttributeEnricher.enrichEvent(null, event);
+
+    when(event.getAttributes()).thenReturn(Attributes.newBuilder().build());
+    when(event.getAttributes().getAttributeMap()).thenReturn(null);
+    resourceAttributeEnricher.enrichEvent(null, event);
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: What are you asserting in this test?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+  public void noResourceInTrace() {
+    // This trace has no resource attributes.
+    StructuredTrace trace = getBigTrace();
+    for (Event event : trace.getEventList()) {
+      int attributeMapSize = 0;
+      if (event.getAttributes() != null && event.getAttributes().getAttributeMap() != null) {
+        attributeMapSize = event.getAttributes().getAttributeMap().size();
+      }
+      resourceAttributeEnricher.enrichEvent(trace, event);
+      if (event.getAttributes() != null && event.getAttributes().getAttributeMap() != null) {
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java:57]: Do you really need these checks if you're constructing the trace yourself?

**Review by suddendust — CHANGES_REQUESTED** [review]
First iteration, please address the changes.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java (hunk 7 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java:57]: Yes. These checks were not before and some of the events of custom trace did not have these attributes which makes sense to test on different edge cases.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java (hunk 6 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: error logs or any exception.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 1 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: fixed

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 2 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: ok

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 3 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: done

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java (hunk 5 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java:44]: Didn't get you, findFirst in the map?

File: .github/workflows/pr-build.yml [hunk 8]
```diff
@@ -1,9 +1,6 @@
 name: build and validate
 on:
   push:
-  pull_request_target:
```
Comment (Sunn-y-Arora) [inline .github/workflows/pr-build.yml:4]: any reason to remove these ?

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.Map;
+import java.util.Optional;
+import org.hypertrace.core.datamodel.AttributeValue;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.traceenricher.enrichment.AbstractTraceEnricher;
+import org.hypertrace.traceenricher.enrichment.clients.ClientRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (Sunn-y-Arora) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java:16]: add a gist of the enricher as comments pls

Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java:16]: sure

File: .github/workflows/pr-build.yml (hunk 8 above)
Comment (suddendust) [inline .github/workflows/pr-build.yml]: Why?

File: .github/workflows/pr-test.yml [hunk 10]
```diff
@@ -1,9 +1,6 @@
 name: test
 on:
   push:
-  pull_request_target:
```
Comment (suddendust) [inline .github/workflows/pr-test.yml]: Same

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  @Override
+  public void init(Config enricherConfig, ClientRegistry clientRegistry) {
+    if (enricherConfig.hasPath(RESOURCE_ATTRIBUTES_CONFIG_KEY)) {
+      resourceAttributesToAdd = enricherConfig.getStringList(RESOURCE_ATTRIBUTES_CONFIG_KEY);
+    }
+  }
+
+  private boolean isValidEvent(Event event) {
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Move below the public method.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    if (resourceAttributesToAdd.isEmpty()) {
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: If there are no resource attributes to add, then remove this enricher from the Enricher DAG? This logic should be moved to when Trace Enricher is bootstrapping.

File: .github/workflows/pr-build.yml (hunk 8 above)
Comment (Harnoor-se7en) [reply .github/workflows/pr-build.yml]: same workflow is running for pull and push. hence 2 workflows running everytime.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      return false;
+    }
+    if (event.getResourceIndex() < 0) {
+      return false;
+    }
+    return (event.getAttributes() != null) && (event.getAttributes().getAttributeMap() != null);
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Inline all of these predicates: `event.getResourceIndex() >= 0 && (event.getAttributes() != null) && (event.getAttributes().getAttributeMap() != null)`

File: .github/workflows/pr-test.yml (hunk 10 above)
Comment (Harnoor-se7en) [reply .github/workflows/pr-test.yml]: same workflow is running for pull and push. hence 2 workflows running everytime.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            getResourceAttribute(trace, event, resourceAttributeKey);
+        resourceAttributeMaybe.ifPresent(
+            attributeValue -> {
+              if (NODE_SELECTOR_KEY.equals(resourceAttributeKey)) {
+                attributeValue.setValue(
+                    attributeValue
+                        .getValue()
+                        .substring(attributeValue.getValue().lastIndexOf('/') + 1));
+              }
+              attributeMap.putIfAbsent(resourceAttributeKey, attributeValue);
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Better:

```
attributeMap.computeIfAbsent(resourceAttributeKey, key -> {
                  if (NODE_SELECTOR_KEY.equals(resourceAttributeKey)) {
                attributeValue.setValue(
                    attributeValue
                        .getValue()
                        .substring(attributeValue.getValue().lastIndexOf('/') + 1));
              }
})
```

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java (hunk 5 above)
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java:44]: Or:
```
attributes.entrySet().stream().filter((k, v) -> k.equals(key)).findFirst()
```

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 13 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: Ok. Though I feel inline predicates makes code a bit unreadable :stuck_out_tongue:

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java (hunk 6 above)
Comment (suddendust) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: Where's the assertion? A test method should only throw an `AssertionError`, as a result of an assertion failing. You can do:

```
try {
    resourceAttributeEnricher.enrichEvent(null, event);
} catch (Exception e) {
    throw new AssertionError("...")
}
```

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    Event event =
+        Event.newBuilder()
+            .setAttributes(attributes)
+            .setEventId(createByteBuffer("event1"))
+            .setCustomerId(TENANT_ID)
+            .build();
+    event.setResourceIndex(0);
+    resourceAttributeEnricher.enrichEvent(structuredTrace, event);
+    assertEquals(
+        resourceAttributeEnricher.getResourceAttributesToAdd().size() - 2,
```
Comment (suddendust) [inline hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: Why is this method in the public API? Seems like it has been added just to facilitate this test.

**Review by suddendust — CHANGES_REQUESTED** [review]
Review iteration 2, yet to review the test cases properly. Will do once these comments are resolved.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 12 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: If we are specifying an empty list then we should not add the enricher in enricherInfoList. So all enrichers are initialized only based on whether the owner is adding it in enriched info config or not. Adding a second check before init seems unnecessary to me. I am removing this condition hence.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java (hunk 14 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricher.java]: done. Added return statement in this block, pls re-check.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java (hunk 5 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java:44]: So are u suggesting - `Optional.ofNullable(attributes).get().getAttributeMap().entrySet().stream().filter(k -> k.equals(key)).map(k -> k.getValue()).findFirst();`? 
I think the current code is also elegant though let me know if there is still a more elegant way.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java (hunk 15 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: Hmm yeah, currently there is no use for this getter. fixing it.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java (hunk 6 above)
Comment (Harnoor-se7en) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/test/java/org/hypertrace/traceenricher/enrichment/enrichers/ResourceAttributeEnricherTest.java]: yeah, so I think this test is useless as I am not throwing any exceptions. removing it then.

File: hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java (hunk 5 above)
Comment (suddendust) [reply hypertrace-trace-enricher/hypertrace-trace-enricher-impl/src/main/java/org/hypertrace/traceenricher/util/EnricherUtil.java:44]: See my latest comment on this. You can stream the map entries, add a filter and then do a findFirst. This returns an Optional by default. Cleaner IMO but this is fine too.

File: .github/workflows/pr-build.yml (hunk 8 above)
Comment (suddendust) [reply .github/workflows/pr-build.yml]: Remove as discussed, same for `pr-test.yml`.

**Review by suddendust — APPROVED** [review]
LGTM
